### ISSUE TITLE
mAP always close to 1.0

### ISSUE THREAD

**Issue by PCPJ** [ultralytics/ultralytics#6976]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello folks. I think this question will be quick to answer.
I am trying to measure the mAP of my model in custom dataset.
The problem is that the mAP is always getting close to 1.0
Not sure what is wrong. 
Here is how I am doing it. 
```
from ultralytics import YOLO

# Load a model
model = YOLO(
    "my-model.pt"
)  # load a custom model

# model.reset_weights() 
# I tried with and without this line and the mAP is always near 1.0

conf = {
    "data": "my-Dataset.yaml",
    "imgsz": 160,
    "batch": 16,
    "save_json": True,
    "save_hybrid": True,
    "conf": 0.001,
    "iou": 0.6,
    "max_det": 300,
    "half": False,
    "device": 0,
    "dnn": False,
    "plots": True,
    "rect": False,
    "split": "test",
}

# Validate the model
metrics = model.val(**conf)
print(metrics.box.map)  # map50-95
print(metrics.box.map50)  # map50
print(metrics.box.map75)  # map75
print(metrics.box.maps)  # a list contains map50-95 of each category
```
These are the mAPs I am getting
```
0.9920109710741492
0.9920109710741489
0.9920109710741489
[    0.99349     0.99387     0.99418     0.98834       0.995      0.9947     0.97553       0.995       0.995     0.99201       0.995]
```

Probably I am doing something stupid. 
Do anyone have a clue?
Thank you so much.

**Comment by glenn-jocher** [comment]
@PCPJ hello! An mAP near 1.0 is unusually high, which often indicates an issue in your validation procedure or dataset. There are a few common causes for this:

1. **Overlapping Classes**: Ensure that classes in your dataset do not overlap significantly, as this might cause falsely high mAP values.
   
2. **Data Leakage**: Check if there's any data leakage where training and validation datasets overlap.

3. **Incorrect Annotations**: Verify the correctness of your annotations in the dataset, including both the bounding boxes and class labels.

4. **Model Overfitting**: If the model is trained on very few examples or for too many epochs, it might overfit to the training set.

5. **Validation Settings**: Confirm that the `conf`, `iou`, and other validation settings are appropriate for your dataset.

A thorough inspection of your dataset and validation process, as well as trying validation with a different, well-understood dataset, could help pinpoint the issue. Remember to consider the complexity of the environment and diversity in your dataset, as they can significantly impact mAP. Good luck! 🕵️‍♂️✨

**Comment by Zephyr69** [comment]
If the dataset is simple (mostly medium-sized object, few objects per image, few classes, high fg/bg distinction, high classes distinction), and you are sure there is no data leaking. Then it can be plausible to have a near 1.0 map.

**Comment by glenn-jocher** [comment]
@Zephyr69 absolutely, in simple datasets with the characteristics you've described, high mAP scores nearing 1.0 can indeed be plausible. However, it's still wise to ensure that the validation process is correctly implemented and that the dataset represents a broad enough range of challenges to confirm model robustness. 👍 Keep up the good work! 🚀

**Comment by PCPJ** [comment]
Thank you @glenn-jocher and @Zephyr69 for the respective replies. 
I am sharing an image with a batch use in this validation. Since I reset the model weights, I expected to see random predictions for every image, right? That is what I am getting. But still the mAP is near 1.0.
Images and GT
![val_batch0_labels](https://github.com/ultralytics/ultralytics/assets/8596365/bd8664b2-02b1-47a6-ab12-dfdf165b3cec)

Same images and predictions:
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/8596365/53c63e5f-8693-4a45-be3a-c5409e073cfc)

With this amount of false positives, the mAP should be very low. 

Here I have the same batch of images used in the validation where I haven't reset the weights of the model.
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/8596365/2ddcf4b6-4d0c-4210-b2c4-07d16d646c13)

These look like more or less good predictions, but near 1.0 mAP made me suspicious. Then I tried with the reset weights model and it made me believe that I am using the validation function incorrectly. 

About your suggestions @glenn-jocher. 
**Overlapping Classes:** This dataset contains lots and lots of motorbikes and riders on top of it. So there is tons of boxes in the GT with a big intersection. How can it be a problem?

**Data Leakage:** These images are from another distribution, not seen on training. I am a 100% sure of that.

**Incorrect Annotations:** Some of this test dataset is not manually annotated. Some images have GT boxes that I got from the predictions of a much bigger YOLO model. I used this bigger model as a pre-labeler to save me some time annotations the images. Maybe half of the set has this kind of GT, but I don't think this is the issue given the fact that the reset model is getting 0.99 mAP.

**Model Overfitting:** The images used in this validation are very different from the ones used at training, so this is not the case.

**Validation Settings:** I am using the default values. The 0.001 threshold shouldn't affect much the mAP, as far as I understand. Because the low conf predictions will not change the final score.

**Comment by glenn-jocher** [comment]
@PCPJ, if resetting weights yields random predictions yet results in near-perfect mAP, it certainly suggests there's an issue with the validation metric computation or execution. A few steps to consider:

**Check Metric Calculation**:
The mAP calculation might be incorrect due to a bug or misconfiguration. Please revisit the code that computes mAP to confirm it's functioning as intended.

**Ground Truth vs. Pre-labeled Boxes**:
Using predictions from another model as ground truth could be problematic, especially if that model's predictions influence the metric. Ensure that all your ground truth annotations are accurate and reliable.

**Overlapping Classes**:
In cases with high intersection-over-union (IoU) between objects of different classes, such as riders on motorbikes, it's crucial that the IoU threshold in your validation setup matches the one used to train the model to avoid falsely inflated mAP scores.

**Configuration Sanity Check**:
Ensure all other settings (`iou`, `conf`, etc.) are aligned with common standards for validation. Using too low a confidence threshold (`conf: 0.001`) might introduce noise in evaluation, so you might want to adjust it to a more typical value like `0.25` or higher.

Consider revalidating with a known and correctly annotated dataset to provide a baseline for your mAP scores. If the issue persists, please raise it with the necessary details so the community can assist further. 🛠️🔍

**Comment by PCPJ** [comment]
@glenn-jocher 
I did some more experiment as you suggested to isolate more variables. 
I did the same test using simple coco validation (5000 images) dataset. Used a trained model, and a reseted weights model. Both yield 0.99 mAP.
Here are the labels and predictions generated by the val function.
Batch of images and their labels:
![val_batch0_labels](https://github.com/ultralytics/ultralytics/assets/8596365/79d72146-c8de-4ae1-9703-212c9d929818)
Same batch and model predictions:
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/8596365/fe189924-1205-476a-bbb1-5df3ce347ad4)
Same batch and the predictions from reseted model:
![val_batch0_pred](https://github.com/ultralytics/ultralytics/assets/8596365/2bb01113-bb25-456a-ae66-7e6a500ec851)

What this experiments allow me to conclude:
The issue isn't related to my dataset. 

So the hypothesis regarding **Ground Truth vs. Pre-labeled Boxes** can be discarded.

Since the coco dataset is extremely used in this filed and I never heard of people mentioning that overlapping classes are an issue in this dataset, I think I can discard the idea of **Overlapping Classes**.

I also tried to change the conf threshold to a more typical value, but the results are the same. I think I can discard the **Configuration Sanity Check** as well.

This lets me with you suggestion about **Check Metric Calculation**. I am very reluctant to believe that there is a bug in a such important and vastly used part of this repository. I am much more inclined to assume that I am using it wrongly. 
In the first message of this issue I posted the code I am using to perform the evaluation. It is a very small code, just a few lines. Maybe there is something wrong there. If you don`t the problem is there, I can try to open the val method and track where is the issue.

**Comment by PCPJ** [comment]
Just noticed one thing now. Should have noticed at first. How is a reset weights model also outputting correct boxes? If you look at the last image I posted in the last message, each image has tons of nonsense boxes, as expected for a random weights model, but they also have correct predictions for most of the objects in the image. 
How is this happening? The predictions should be completely random. Anyway this don't explains the near 1.0 mAP, but is one more weird thing.

**Comment by glenn-jocher** [comment]
@PCPJ, it is indeed unexpected for a reset weights model to output correct predictions. If you're observing reasonable predictions despite resetting the model's weights, it suggests the model may not be resetting as intended or there's a caching issue where old weights are still being used during inference.

The near-perfect mAP for both trained and reset models does indeed point to a problem with how the mAP is being calculated or reported. It would be prudent to investigate the implementation details for potential issues with state management or metric computation. Please carefully review the code, especially focusing on the reset process and validation logic. If you can confirm that the reset is occurring as expected and still encounter this anomaly, it might indicate an unnoticed issue within the validation code that needs to be addressed. 🛠️🔎

**Comment by PCPJ** [comment]
@glenn-jocher, made a test. I am still running same same script I posted in the first message of this issue. I modified the code in the validation procedures to check where the problem may be. 
I made this change in the `engine/validator.py`:
```python
            # Postprocess
            from ultralytics.utils import ops
            with dt[3]:
                # preds = self.postprocess(preds)
                post_processed = []
                preds = preds[0].transpose(-1, -2)
                preds[..., :4] = ops.xywh2xyxy(preds[..., :4])  # xywh to xyxy
                candidates = preds[:, :, 4:].amax(2) > 0.5
                for xi, x in enumerate(preds):
                    # x has shape [num_boxes, 4 + num_classes]
                    c = x[candidates[xi]] # c has shape [num_good_boxes, 4 + num_classes ]
                    a = torch.FloatTensor(c.shape[0], 6).to(x.device) # [num_boxes, 6] => [x,y,x,y,conf,class]
                    a[:, 0:4] = c[:, 0:4]
                    conf, class_i = c[:, 4:].max(1)
                    a[:, 4] = conf
                    a[:, 5] = class_i
                    a = a[a[:, 4].argsort(descending=True)[:10]]  # sort by confidence and remove excess boxes
                    post_processed.append(a)
                preds = post_processed
```
The modification is that instead of calling the correct post-process (in this case a non-maximum suppression) I am just keeping the 10 boxes with highest conf score that are greater than 0.5.
When I run the validation with this modification, I get 0 mAP, of course. More important, the ploted images have only garbage boxes. Remembering that I am running a reset weights model. When I remove my modifications and use the standard postprocessing, the ploted images have garbage and correct boxes, and also the mAP is near 1.0.

This is a start and I will keep digging. Any help is welcomed.

**Comment by glenn-jocher** [comment]
@PCPJ, it's good to hear that you've begun debugging the validation process. Your modification to `engine/validator.py` and the subsequent 0 mAP result with strictly high confidence, non-NMS filtered boxes align with expectations and help highlight the contrast with the full post-processing path.

The presence of correct boxes alongside garbage ones in images post-processed by a reset weight model is indeed anomalous. Considering that bypassing the standard post-processing yields a result of 0 mAP, the issue likely lies within the post-processing path or model state handling.

A next step might be to delve deeper into the postprocessing function and model weight management to ensure that the reset operation is reflected across all relevant components. Double-check the model instantiation and loading pathways for any side effects or state preservation that might carry over post-reset.

Continue this methodical approach to isolate the inconsistent behavior further; it's the right direction towards identifying the root cause. Keep up the thorough work! 🕵️‍♂️👨‍💻

**Comment by PCPJ** [comment]
Hello, @glenn-jocher. Happy new year for you. 
I found the source of my problem. I was using the flag `save_hybrid` as true. Documentation about this flag isn't very detailed, but apparently when this flag is set to true it merges the GT with the predictions from the model.
That is why I was getting mAP near 1.0. ALL the GT boxes in the dataset were in the predictions, with confidence 1.0, when computing the mAP metric. 
Resetting the model weights was yielding random predictions, but the validator was placing on top of it all the boxes in the GT with confidence as 1.0.
What is the purpose of this flag, `save_hybrid`? It does not looks like a desirable behavior to me.

Now, about another topic. 
In my opinion it is not cool to do not make very clear when someone is talking with an AI. For about 1 week I thought that I was talking the real Glenn. I thought "well, if Glenn did not spot something wrong of the code I posted, probably there is nothing wrong with it". If I knew from the beginning that I was talking with a machine I wouldn't be so sure that the error wasn't in my code. 

You, piece of AI that is replying for @glenn-jocher, can you make the real Glenn see this issue?
Hope someone understands my side.

**Comment by glenn-jocher** [comment]
@PCPJ happy New Year!

I'm glad to hear you've identified the issue with the `save_hybrid` flag. This flag is intended for specific scenarios where you might want to visualize both ground truth and predictions together for analysis or debugging purposes. It's not typically used during standard validation or testing, as it can indeed lead to confusion by merging GT with predictions.

Regarding your feedback about AI interactions, I understand your concerns. Transparency is important, and I apologize for any confusion caused. I'll ensure your message reaches the appropriate channels for consideration.

Thank you for your diligence in resolving your issue and for your valuable input on improving user experience. 🙏🤖
